events/events.go

```diff
@@ -165,6 +172,57 @@ type XRPCStreamEvent struct {
 	PrivUid         models.Uid `json:"-" cborgen:"-"`
 	PrivPdsId       uint       `json:"-" cborgen:"-"`
 	PrivRelevantPds []uint     `json:"-" cborgen:"-"`
+	Preserialized   []byte     `json:"-" cborgen:"-"`
+}
+
+func (evt *XRPCStreamEvent) Serialize(wc io.Writer) error {
+	header := EventHeader{Op: EvtKindMessage}
+	var obj lexutil.CBOR
+
+	switch {
+	case evt.Error != nil:
+		header.Op = EvtKindErrorFrame
+		obj = evt.Error
+	case evt.RepoCommit != nil:
+		header.MsgType = "#commit"
+		obj = evt.RepoCommit
+	case evt.RepoHandle != nil:
+		header.MsgType = "#handle"
+		obj = evt.RepoHandle
+	case evt.RepoIdentity != nil:
+		header.MsgType = "#identity"
+		obj = evt.RepoIdentity
+	case evt.RepoAccount != nil:
+		header.MsgType = "#account"
+		obj = evt.RepoAccount
+	case evt.RepoInfo != nil:
+		header.MsgType = "#info"
+		obj = evt.RepoInfo
+	case evt.RepoMigrate != nil:
+		header.MsgType = "#migrate"
+		obj = evt.RepoMigrate
+	case evt.RepoTombstone != nil:
+		header.MsgType = "#tombstone"
+		obj = evt.RepoTombstone
+	default:
+		return fmt.Errorf("unrecognized event kind")
+	}
+
+	cborWriter := cbg.NewCborWriter(wc)
+	if err := header.MarshalCBOR(cborWriter); err != nil {
+		return fmt.Errorf("failed to write header: %w", err)
+	}
+	return obj.MarshalCBOR(cborWriter)
+}
+
+// serialize content into Preserialized cache
+func (evt *XRPCStreamEvent) Preserialize() {
+	if evt.Preserialized != nil {
+		return
+	}
+	var buf bytes.Buffer
+	evt.Serialize(&buf)
```

Comment:
probably still want to check  (or explicitly ignore if we are extremely confident) errors